internal/bson2/logging.go

```diff
@@ -0,0 +1,273 @@
+// for example, both int32 and int64 values are returned with [slog.KindInt64],
+// arrays are treated as documents, and empty documents are omitted.
+// More information is subsequently lost in handlers output;
+// for example, float64(42), int32(42), and int64(42) values would all look the same
+// (`f64=42 i32=42 i64=42` or `{"f64":42,"i32":42,"i64":42}`).
+func slogValue(v any) slog.Value {
+	switch v := v.(type) {
+	case *Document:
+		var attrs []slog.Attr
+
+		for _, f := range v.fields {
+			attrs = append(attrs, slog.Attr{Key: f.name, Value: slogValue(f.value)})
+		}
+
+		return slog.GroupValue(attrs...)
+
+	case RawDocument:
+		if v == nil {
+			return slog.StringValue("RawDocument<nil>")
+		}
+
+		return slog.StringValue("RawDocument<" + strconv.Itoa(len(v)) + ">")
+
+	case *Array:
+		var attrs []slog.Attr
+
+		for i, v := range v.elements {
+			attrs = append(attrs, slog.Attr{Key: strconv.Itoa(i), Value: slogValue(v)})
+		}
+
+		return slog.GroupValue(attrs...)
+
+	case RawArray:
+		if v == nil {
+			return slog.StringValue("RawArray<nil>")
+		}
+
+		return slog.StringValue("RawArray<" + strconv.Itoa(len(v)) + ">")
+
+	case float64:
+		// for JSON handler to work
+		switch {
+		case math.IsNaN(v):
+			return slog.StringValue("NaN")
+		case math.IsInf(v, 1):
+			return slog.StringValue("+Inf")
+		case math.IsInf(v, -1):
+			return slog.StringValue("-Inf")
+		}
+
+		return slog.Float64Value(v)
+
+	case string:
+		return slog.StringValue(v)
+
+	case Binary:
+		return slog.StringValue(fmt.Sprintf("%#v", v))
+
+	case ObjectID:
+		return slog.StringValue("ObjectID(" + hex.EncodeToString(v[:]) + ")")
+
+	case bool:
+		return slog.BoolValue(v)
+
+	case time.Time:
+		return slog.TimeValue(v.Truncate(time.Millisecond).UTC())
+
+	case NullType:
+		return slog.Value{}
+
+	case Regex:
+		return slog.StringValue(fmt.Sprintf("%#v", v))
+
+	case int32:
+		return slog.Int64Value(int64(v))
+
+	case Timestamp:
+		return slog.StringValue(fmt.Sprintf("%#v", v))
+
+	case int64:
+		return slog.Int64Value(v)
+
+	default:
+		panic(fmt.Sprintf("invalid BSON type %T", v))
+	}
+}
+
+// logMessage returns an indented representation of any BSON value as a string,
+// somewhat similar (but not identical) to JSON or Go syntax.
+// It may change over time.
+//
+// The result is optimized for large values such as full request documents.
+// All information is preserved.
+func logMessage(v any) string {
+	return logMessageIndent(v, "")
+}
+
+// logMessageIndent is a variant of [logMessage] with an indentation for recursive calls.
+func logMessageIndent(v any, indent string) string {
+	switch v := v.(type) {
+	case *Document:
+		l := len(v.fields)
+		if l == 0 {
+			return "{}"
+		}
+
+		if flowLimit > 0 {
+			res := "{"
+
+			for i, f := range v.fields {
+				res += strconv.Quote(f.name) + `: `
+				res += logMessageIndent(f.value, "")
+
+				if i != l-1 {
+					res += ", "
+				}
+			}
+
+			res += `}`
+
+			if len(res) < flowLimit {
+				return res
+			}
```

Comment:
If `res` exceeds `flowLimit`, what should it return? It seems to behave same as flowLimit==0.

Reply:
Well, yeah, that's the point of that code. Or I don't understand the question

Reply:
Okay just checking I understood the expectation :+1: